File: shared/src/commonMain/kotlin/pe/ecolecta/domain/usecase/entrega/RegistrarLoteUseCase.kt
package pe.ecolecta.domain.usecase.entrega

import pe.ecolecta.domain.model.AccionAuditoria
import pe.ecolecta.domain.DeviceIdProvider
import pe.ecolecta.domain.EntregaInvalidaException
import pe.ecolecta.domain.Reloj
import pe.ecolecta.domain.model.Auditoria
import pe.ecolecta.domain.model.Entrega
import pe.ecolecta.domain.model.SyncState
import pe.ecolecta.domain.nuevoId
import pe.ecolecta.domain.repository.EntregaRepository
import pe.ecolecta.domain.repository.ProveedorRepository

data class ItemLote(val proveedorId: String, val litros: Double, val tachos: Int, val observaciones: String? = null)

/** Todas las entregas del lote comparten un lote_id y se guardan en una única transacción: si una falla, ninguna queda (§26). */
class RegistrarLoteUseCase(
    private val entregaRepository: EntregaRepository,
    private val proveedorRepository: ProveedorRepository,
    private val reloj: Reloj,
    private val deviceIdProvider: DeviceIdProvider,
) {
    suspend operator fun invoke(
        jornadaId: String,
        usuarioId: String,
        zonaId: String,
        vehiculoId: String,
        items: List<ItemLote>,
    ): Result<List<Entrega>> {
        if (items.isEmpty()) return Result.failure(EntregaInvalidaException.LoteVacio)

        val loteId = nuevoId()
        val ahora = reloj.ahora().toEpochMilliseconds()
        val deviceId = deviceIdProvider.obtenerId()

        val entregas = mutableListOf<Entrega>()
        val auditorias = mutableListOf<Auditoria>()

        for (item in items) {
            val proveedor = proveedorRepository.obtenerPorId(item.proveedorId)
                ?: return Result.failure(IllegalStateException("Proveedor no encontrado"))
            if (item.litros > proveedor.capacidadTotalL) {
                return Result.failure(EntregaInvalidaException.SuperaCapacidad(proveedor.capacidadTotalL))
            }

            val entrega = Entrega.crear(
                id = nuevoId(),
                jornadaId = jornadaId,
                proveedorId = item.proveedorId,
                usuarioId = usuarioId,
                zonaId = zonaId,
                vehiculoId = vehiculoId,
                litros = item.litros,
                tachos = item.tachos,
                observaciones = item.observaciones,
                registradoEn = ahora,
                deviceId = deviceId,
                loteId = loteId,
            ).getOrElse { return Result.failure(it) }

            entregas += entrega
            auditorias += Auditoria(
                id = nuevoId(),
                entidad = "entrega",
                entidadId = entrega.id,
                accion = AccionAuditoria.CREAR,
                valorAntes = null,
                valorDespues = "litros=${item.litros};tachos=${item.tachos}",
                motivo = null,
                usuarioId = usuarioId,
                ocurridoEn = ahora,
                deviceId = deviceId,
                syncState = SyncState.PENDING,
            )
        }

        return runCatching {
            entregaRepository.registrarLote(entregas, auditorias)
            entregas
        }
    }
}

File: web/app/Application/Entregas/RegistrarLoteUseCase.php
<?php

namespace App\Application\Entregas;

use App\Domain\Auditoria\AccionAuditoria;
use App\Domain\Auditoria\Auditoria;
use App\Domain\Entregas\Entrega;
use App\Domain\Entregas\EntregaRepositoryInterface;
use App\Domain\Entregas\Exceptions\EntregaInvalidaException;
use App\Domain\Proveedores\EstadoProveedor;
use App\Domain\Proveedores\ProveedorRepositoryInterface;
use DateTimeImmutable;
use Illuminate\Support\Str;
use RuntimeException;

/** Todas las entregas del lote comparten un lote_id y se guardan en una única transacción: si una falla, ninguna queda. */
final class RegistrarLoteUseCase
{
    public function __construct(
        private readonly EntregaRepositoryInterface $entregas,
        private readonly ProveedorRepositoryInterface $proveedores,
    ) {}

    /**
     * @param  list<ItemLote>  $items
     * @return list<Entrega>
     */
    public function ejecutar(int $jornadaId, int $usuarioId, int $zonaId, int $vehiculoId, array $items): array
    {
        if ($items === []) {
            throw EntregaInvalidaException::loteVacio();
        }

        $loteId = (string) Str::uuid();
        $ahora = new DateTimeImmutable;

        $entregas = [];
        $auditorias = [];

        foreach ($items as $item) {
            $proveedor = $this->proveedores->buscarPorId($item->proveedorId);

            if ($proveedor === null) {
                throw new RuntimeException('Proveedor no encontrado.');
            }

            if ($proveedor->estado !== EstadoProveedor::Activo) {
                throw EntregaInvalidaException::proveedorNoActivo($proveedor->estado);
            }

            if ($item->litros > $proveedor->capacidadTotalL()) {
                throw EntregaInvalidaException::superaCapacidad($proveedor->capacidadTotalL());
            }

            $entregas[] = Entrega::crear(
                jornadaId: $jornadaId,
                proveedorId: $item->proveedorId,
                usuarioId: $usuarioId,
                zonaId: $zonaId,
                vehiculoId: $vehiculoId,
                litros: $item->litros,
                tachos: $item->tachos,
                observaciones: $item->observaciones,
                registradoEn: $ahora,
                loteId: $loteId,
            );

            $auditorias[] = new Auditoria(
                id: null,
                entidad: 'entrega',
                entidadId: 0,
                accion: AccionAuditoria::Crear,
                valorAntes: null,
                valorDespues: "litros={$item->litros};tachos={$item->tachos}",
                motivo: null,
                usuarioId: $usuarioId,
                ocurridoEn: $ahora,
            );
        }

        return $this->entregas->registrarLote($entregas, $auditorias);
    }
}
